plugins/nf-google/src/main/nextflow/cloud/google/batch/client/BatchClient.groovy

```diff
@@ -33,160 +38,73 @@ class BatchClient {
 
     protected String projectId
     protected String location
-    protected GoogleCredentials credentials
-    protected String BATCH_ENDPOINT = 'https://batch.googleapis.com/v1'
-    protected String LOGS_ENDPOINT = 'https://logging.googleapis.com'
+    protected BatchServiceClient batchServiceClient
 
     BatchClient(BatchConfig config) {
         this.projectId = config.googleOpts.projectId
         this.location = config.googleOpts.location
-        this.credentials = config.credentials
+        this.batchServiceClient = createBatchService(config)
     }
 
     /** Only for testing - do not use */
     protected BatchClient() {}
 
-    protected String getToken() {
-        // only for debugging purpose
-        if( System.getenv('_GOOGLE_ACCESS_TOKEN') ) {
-            return System.getenv('_GOOGLE_ACCESS_TOKEN')
+    protected CredentialsProvider createCredentialsProvider(BatchConfig config) {
+        if( !config.getCredentials() )
+            return null
+        return new CredentialsProvider() {
+            @Override
+            Credentials getCredentials() throws IOException {
+                return config.getCredentials()
+            }
         }
-        // use native API
-        credentials.refreshIfExpired()
-        final result = credentials.getAccessToken()?.getTokenValue()
-        return result
-    }
-
-    BatchJsonResponse submitJob(String jobId, BatchJob request) {
-        final action = "$BATCH_ENDPOINT/projects/${projectId}/locations/${location}/jobs?job_id=${jobId}"
-        final resp = post(action, request.toJson())
-        trace('POST', action, resp.text)
-        return new BatchJsonResponse(resp.text)
     }
 
-    BatchJsonResponse describeJob(String jobId) {
-        final action = "$BATCH_ENDPOINT/projects/${projectId}/locations/$location/jobs/$jobId"
-        final resp = get(action)
-        trace('GET', action, resp.text)
-        return new BatchJsonResponse(resp.text)
+    protected BatchServiceClient createBatchService(BatchConfig config) {
+        final provider = createCredentialsProvider(config)
+        if( provider ) {
+            log.debug "Creating Batch service client with config credentials"
+            final settings = BatchServiceSettings.newBuilder().setCredentialsProvider(provider).build()
+            return BatchServiceClient.create(settings)
```

Comment:
Have you tried by setting the cred file via the nextflow file instead of the env variable to validate this piece of code?

Reply:
I didn't know the creds file can be specified in the nextflow config? I use the env variable and I set `google.project` and `google.location` in my config.

Reply:
ok, this looks fine